### PR TITLE
#176 - Fix dynamic tuple encoding

### PR DESCRIPTION
**Pull request by biboran**
As per #176 
* Added `isDynamic` and `headsCount` properties to the `ABIEncodedParameter` to address encoding cases where tuple is dynamic
* Split a part of the tuple encoding functionality into `ABITupleEncoding` class. I am not sure if it's a good naming
* Added a sample 0x contract call encoding

As a side effect of this change you no longer have to worry about unwrapping enveloped tuples as is illustrated in the `testTwoAddressFixedArraysOfTwo` test

### CODE REVIEW

**Comment by rockfridrich** [comment]
@biboran thank you, the new implementation looks excellent! 
I will check it later this week. 
Maybe we could rename `ABITupleEncoding` to `ABIEncodedTuple`. It will make it more readable from my perspective. 
Also, I will try to guess if it is possible to get rid of checking `isDynamic()` in many functions.

Should we try to write tests for such 0x function? It consists of a dynamic collection of dynamic tuples  
```js
/// @dev Synchronously executes multiple calls of fillOrder.
/// @param orders Array of order specifications.
/// @param takerAssetFillAmounts Array of desired amounts of takerAsset to sell in orders.
/// @param signatures Proofs that orders have been created by makers.
/// @return Amounts filled and fees paid by makers and taker.
///         NOTE: makerAssetFilledAmount and takerAssetFilledAmount may include amounts filled of different assets.
function batchFillOrders(
    Order[] memory orders,
    uint256[] memory takerAssetFillAmounts,
    bytes[] memory signatures
)
    public;
    returns (FillResults memory totalFillResults)
```

**Comment by biboran** [comment]
> Maybe we could rename `ABITupleEncoding` to `ABIEncodedTuple`. It will make it more readable from my perspective.

I am not sure naming it `ABIEncodedTuple` will help since we already have `EncodedABITuple`

**Comment by biboran** [comment]
> Should we try to write tests for such 0x function? It consists of a dynamic collection of dynamic tuples

We desperately need more unit tests for ABI encoding. The existing unit tests were very helpful in regression testing. We should implement more ABI encoding unit tests and ganache/infura integration tests. 
Lets also split ganache/infura test into a separate target

**Comment by rockfridrich** [comment]
@biboran do you have an idea how to make ganache run as a part of the testing pipeline?

File: Web3Swift/ABI/Parameters/ABITupleEncoding.swift [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            let parameterTails = try parameter.tails(
+                offset: additionalOffset
+            )
+            tails += parameterTails
+            additionalOffset += parameterTails.count
+        }
+        return heads + tails
+    }
+    
+    //TODO: headsCount should probably be injected
```
Comment (rockfridrich) [inline Web3Swift/ABI/Parameters/ABITupleEncoding.swift]: @biboran in which case we need to inject them? If we want manually construct data?

Comment (biboran) [reply Web3Swift/ABI/Parameters/ABITupleEncoding.swift]: `headsCount` function is not a part of any protocol in this class so it should probably be injected from the outside and removed from this class. For example we could make a separate `HeadsCount: IntegerScalar` class

**Comment by biboran** [comment]
> do you have an idea how to make ganache run as a part of the testing pipeline?

I think we can try making a tests runner app for mac os that will spawn ganache and tests [Process](https://developer.apple.com/documentation/foundation/process)

**Comment by biboran** [comment]
Nvmind I found an easy alternative. Lets discuss it in #173

**Comment by rockfridrich** [comment]
@biboran lets merge it and distribute it to CocoaPods. I will spend some time next week to write additional test cases and verify them with `web3.js`. 
Do we need to update docs in `README.md`?

**Comment by biboran** [comment]
I don't think there is any impact on the README

**Pull request merged**